src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,323 @@
+    
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  VARIABLES  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+     /// @notice List of supported ERC20 asset contracts.
+    IERC20[] public assets;
+
+    mapping(IERC20 => AssetData) public _assetData;
+
+    bool public actionsPaused;
+
+    IRateProvider public rateProvider;
+    IEigenStrategyManager public eigenStrategyManager;
+    IynEigen ynEigen;
+
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  INITIALIZATION  ----------------------------------
+    //--------------------------------------------------------------------------------------
+
+    constructor() {
+       _disableInitializers();
+    }
+
+    struct Init {
+        IERC20[] assets;
+        IRateProvider rateProvider;
+        IEigenStrategyManager eigenStrategyManager;
+        IynEigen ynEigen;
+        address admin;
+        address pauser;
+        address unpauser;
+    }
+
+    function initialize(Init calldata init)
+        public
+        notZeroAddress(address(init.rateProvider))
+        notZeroAddress(address(init.admin))
+        notZeroAddress(address(init.pauser))
+        notZeroAddress(address(init.unpauser))
+        initializer {
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, init.admin);
+        _grantRole(PAUSER_ROLE, init.pauser);
+        _grantRole(UNPAUSER_ROLE, init.unpauser);
+
+        for (uint256 i = 0; i < init.assets.length; i++) {
+            if (address(init.assets[i]) == address(0)) {
+                revert ZeroAddress();
+            }
+            assets.push(init.assets[i]);
+            _assetData[init.assets[i]] = AssetData({
+                active: true
+            });
+        }
+
+        rateProvider = init.rateProvider;
+        eigenStrategyManager = init.eigenStrategyManager;
+        ynEigen = init.ynEigen;
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ASSET MANAGEMENT  --------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice Adds a new asset to the system.
+     * @dev Adds an asset to the _assetData mapping and sets it as active. This function can only be called by the strategy manager.
+     * @param asset The address of the ERC20 token to be added.
+     */
+    function addAsset(IERC20 asset) public onlyRole(ASSET_MANAGER_ROLE) whenNotPaused {
+        require(!_assetData[asset].active, "Asset already active");
+
+        assets.push(asset);
+
+        _assetData[asset] = AssetData({
```

Comment:
could we get away with a simple mapping to boolean here?

Reply:
Yes, Reason i did this is in case we add future fields in the future.

If you create a raw boolean mapping, when you add a new field in the future you need to create a new mapping with a struct or single field.

With a struct, you can keep appending fields to the struct, and no new variables need to declared